src/Microsoft.ML.Core/Data/ColumnType.cs

```diff
@@ -886,60 +883,48 @@ private static Type GetRawType(PrimitiveType itemType)
             return typeof(VBuffer<>).MakeGenericType(itemType.RawType);
         }
 
-        private static int ComputeSizes(int[] dims, out int[] sizes)
+        private static int ComputeSize(ImmutableArray<int> dims)
         {
-            sizes = new int[dims.Length];
             int size = 1;
-            for (int i = dims.Length; --i >= 0; )
-                size = sizes[i] = checked(size * dims[i]);
+            for (int i = 0; i < dims.Length; ++i)
+                size = checked(size * dims[i]);
             return size;
         }
 
-        public int DimCount { get { return _dims != null ? _dims.Length : _size > 0 ? 1 : 0; } }
```

Comment:
This is also potentially controversial; this was not part of the original review comments but it was so bad I felt compelled to improve it. I've replaced all this machinery around `_dims` and `DimCount` and `GetDim` with a single `ImmutableArray<int> Dimensions`. (In general I think it would be positive to start using this library. Roslyn uses it a lot and we have a similar desire for immutable collections of numbers.)

`DimCount` was an incredibly odd structure, reflecting `_dims`. Let's imagine you created something with dims `[2, 5]`, `[7]`, and `[0, 5]`. In those cases, the size would be `10`, `7`, `0` respectively, and the dimensions would be equal to what you had passed in. So far so good. But if you passed in `[0]`, the number of dimensions would equal 0, not 1. This was confusing. Also of course we have this thing that is kind of a "list light" because we could not expose `_dims` directly, but neither I guess did we want to expose some sort of .NET type like `IReadOnlyList` or somesuch, so we had this `DimCount` and `GetDim`. Exposing this other structure makes a lot more sense to me.